% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/generics.R, R/celltype.R
\name{AnnotateCellType}
\alias{AnnotateCellType}
\alias{AnnotateCellType.scCATCH}
\alias{AnnotateCellType.SingleR}
\alias{AnnotateCellType.CellMarker}
\alias{AnnotateCellType.SelfMarker}
\title{Cell Type Annotation}
\usage{
AnnotateCellType(input, outdir, project, used = "SingleR", ...)

\method{AnnotateCellType}{scCATCH}(
  input,
  outdir,
  project,
  used,
  species = "Human",
  tissue = NULL,
  strict = FALSE,
  geneinfo = scCATCH::geneinfo,
  cellmatch = scCATCH::cellmatch,
  cell_min.pct = 0.25,
  logfc = 0.25,
  pvalue = 0.05,
  plot.features = c("nFeature_RNA", "percent.mt", "percent.rb")
)

\method{AnnotateCellType}{SingleR}(
  input,
  outdir,
  project,
  used,
  reference = NULL,
  level = NULL,
  species = "Human",
  plot.features = c("nFeature_RNA", "percent.mt", "percent.rb")
)

\method{AnnotateCellType}{CellMarker}(
  input,
  outdir,
  project,
  used,
  species = "Human",
  tissue = NULL
)

\method{AnnotateCellType}{SelfMarker}(input, outdir, project, used, marker.file = NULL)
}
\arguments{
\item{input}{Seurat input object}

\item{outdir}{The ouput directory for rds and plot}

\item{used}{Methods for cell type annotation: scCATCH, SingleR}

\item{species}{Value: Human or 'Mouse'}

\item{tissue}{tissue name in cellmarker database}

\item{strict}{Use the most strict condition to identify marker genes, default is FALSE}

\item{geneinfo}{NCBI gene info for a given species, see \code{\link[scCATCH:demo_geneinfo]{scCATCH::demo_geneinfo()}}}

\item{cellmatch}{Known markers of human and mouse, other species see \code{\link[scCATCH:demo_marker]{scCATCH::demo_marker()}}}

\item{logfc}{Include the gene with at least this fold change of average gene expression compared to every other clusters, default is 0.25}

\item{pvalue}{Include the significantly highly expressed gene with this cutoff of p value from wilcox test compared to every other clusters, default is 0.05}

\item{reference}{Reference dataset in celldex,
HumanPrimaryCellAtlasData、BlueprintEncodeData、MouseRNAseqData、
DatabaseImmuneCellExpressionData、NovershternHematopoieticData、MonacoImmuneData}

\item{level}{Label levels: main (broad), fine (fine-grained), ont (standard in Cell Ontology)}

\item{marker.file}{csv file, including cell_name,marker_genes}

\item{cell_min_pct}{Include the gene detected in at least this many cells in each cluster, default: 0.25}
}
\value{
Return a combined object
}
\description{
Cell Type Annotation
}
\section{scCATCH}{

Cluster-based Toolkit for Cellular Heterogeneity (scCATCH)
from cluster potential marker genes identification to cluster annotation
based on evidence-based score by matching the potential marker genes
with known cell markers in tissue-specific cell taxonomy reference database (CellMatch).

Support species is 'Human' or 'Mouse'
\itemize{
\item source code: \url{https://github.com/ZJUFanLab/scCATCH}
\item quick start: \url{https://raw.githack.com/ZJUFanLab/scCATCH/master/vignettes/tutorial.html}
}
}

\section{SingleR}{

\itemize{
\item source code: \url{https://github.com/dviraran/SingleR}
\item quick start: \url{https://bioconductor.org/packages/devel/bioc/vignettes/SingleR/inst/doc/SingleR.html}
}
\subsection{Two mode:}{
\itemize{
\item use built-in reference: HumanPrimaryCellAtlasData、BlueprintEncodeData、MouseRNAseqData、
DatabaseImmuneCellExpressionData、NovershternHematopoieticData、MonacoImmuneData
see details in \code{browseVignettes("celldex")} or https://bioconductor.org/packages/release/data/experiment/vignettes/celldex/inst/doc/userguide.html#2_General-purpose_references
\item use pre-labelled dataset to annotate: (todo)
See \code{browseVignettes("SingleR::trainSingleR")}
}
}

\subsection{Paramenters}{
\itemize{
\item input:
}
}
}

\section{CellMarker}{

\itemize{
\item homepage: \url{http://bio-bigdata.hrbmu.edu.cn/CellMarker/index.html}
}
}

\section{SelfMarker}{

\itemize{
\item homepage: \url{http://bio-bigdata.hrbmu.edu.cn/CellMarker/index.html}
}
}

\concept{cell type annotation}
